stor/dx.py

```diff
@@ -113,96 +165,727 @@ def temp_url(self, lifetime=300, filename=None):
                 URL will be valid
             filename (str, optional): A urlencoded filename to use for
                 attachment, otherwise defaults to object name
+
+        Raises:
+            ValueError: The path points to a project
+            NotFoundError: The path could not be resolved to a file
         """
-        try:
-            if self.canonical_resource:
-                file_handler = dxpy.DXFile(self.canonical_resource)
-                return file_handler.get_download_url(
-                    duration=lifetime,
-                    preauthenticated=True,
-                    filename=filename,
-                    project=self.canonical_project
-                )[0]
-            else:
-                raise DXError('DX Projects cannot have a temporary download url')
-        except ValueError:
-            raise DXError('DXPaths ending in folders cannot have a temporary download url')
+        if not self.resource:
+            raise ValueError('DX Projects cannot have a temporary download url')
+        with _propagate_dx_exceptions():
+            file_handler = dxpy.DXFile(dxid=self.canonical_resource,
+                                       project=self.canonical_project)
+            return file_handler.get_download_url(
+                duration=lifetime,
+                preauthenticated=True,
+                filename=filename,
+                project=self.canonical_project
+            )[0]
 
     @property
     def resource(self):
+        """Returns the resource as a ``PosixPath`` object or None.
+
+        The returned resource is defined relative to a project.
+        For virtual paths like DXPath('dx://project:dir/file'), the resource
+        is PosixPath('/dir/file').
+        For canonical paths like
+        DXPath('dx://project-FJq288j0GPJbFJZX43fV5YP1:file-FKzjvV80jfv8pFzQ0KfbYY6B'),
+        the resource is PosixPath('/file-FKzjvV80jfv8pFzQ0KfbYY6B').
+
+        The operation parses the object string, and hence no API request is performed.
+        """
         parts = self._get_parts()
         joined_resource = '/'.join(parts[1:]) if len(parts) > 1 else None
         return self.parts_class('/'+joined_resource) if joined_resource else None
 
     def dirname(self):
+        """Returns directory name of path. Returns self if path is a project.
+
+        For canonical paths, returns the project ID, since this parses the object
```

Comment:
`parses the object as string` doesn't really make sense. `To avoid making API calls, canonical paths will return the project ID`

Reply:
Changed